diff -U 1 -r binutils-2.14.90.0.6.orig/opcodes/Makefile.am binutils-2.14.90.0.6/opcodes/Makefile.am
--- binutils-2.14.90.0.6.orig/opcodes/Makefile.am	2004-12-01 15:45:34.068706440 -0500
+++ binutils-2.14.90.0.6/opcodes/Makefile.am	2004-12-01 15:45:43.786229152 -0500
@@ -285,3 +285,3 @@
 libopcodes_la_DEPENDENCIES = $(OFILES) ../bfd/libbfd.la
-libopcodes_la_LIBADD = $(OFILES) @WIN32LIBADD@ ../bfd/libbfd.la
+libopcodes_la_LIBADD = $(OFILES) @WIN32LIBADD@ -L../bfd/.libs ../bfd/libbfd.la
 libopcodes_la_LDFLAGS = -release $(VERSION) @WIN32LDFLAGS@
diff -U 1 -r binutils-2.14.90.0.6.orig/opcodes/Makefile.in binutils-2.14.90.0.6/opcodes/Makefile.in
--- binutils-2.14.90.0.6.orig/opcodes/Makefile.in	2004-12-01 15:45:34.068706440 -0500
+++ binutils-2.14.90.0.6/opcodes/Makefile.in	2004-12-01 15:45:43.800227024 -0500
@@ -395,3 +395,3 @@
 libopcodes_la_DEPENDENCIES = $(OFILES) ../bfd/libbfd.la
-libopcodes_la_LIBADD = $(OFILES) @WIN32LIBADD@ ../bfd/libbfd.la
+libopcodes_la_LIBADD = $(OFILES) @WIN32LIBADD@ -L../bfd/.libs ../bfd/libbfd.la
 libopcodes_la_LDFLAGS = -release $(VERSION) @WIN32LDFLAGS@
