generate more pie scripts

--- ld/emulparams/hppa64linux.sh
+++ ld/emulparams/hppa64linux.sh
@@ -13,6 +13,7 @@
 ENTRY="main"
 TEMPLATE_NAME=elf32
 GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
 
 # We really want multiple .stub sections, one for each input .text section,
 # but for now this is good enough.
--- ld/emulparams/hppalinux.sh
+++ ld/emulparams/hppalinux.sh
@@ -17,6 +17,6 @@
 DATA_PLT=
 PLT_BEFORE_GOT=
 GENERATE_SHLIB_SCRIPT=yes
-oGENERATE_PIE_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
 TEMPLATE_NAME=elf32
 EXTRA_EM_FILE=hppaelf
--- ld/emulparams/elf32bmip.sh
+++ ld/emulparams/elf32bmip.sh
@@ -39,3 +39,4 @@
 MACHINE=
 TEMPLATE_NAME=elf32
 GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
