generate more pie scripts

--- ld/emulparams/hppalinux.sh
+++ ld/emulparams/hppalinux.sh
@@ -17,6 +17,6 @@
 DATA_PLT=
 PLT_BEFORE_GOT=
 GENERATE_SHLIB_SCRIPT=yes
-oGENERATE_PIE_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
 TEMPLATE_NAME=elf32
 EXTRA_EM_FILE=hppaelf
--- ld/emulparams/elf32bmip.sh
+++ ld/emulparams/elf32bmip.sh
@@ -39,3 +39,4 @@
 MACHINE=
 TEMPLATE_NAME=elf32
 GENERATE_SHLIB_SCRIPT=yes
+GENERATE_PIE_SCRIPT=yes
