--- binutils/ld/ldmain.c
+++ binutils/ld/ldmain.c
@@ -306,7 +306,9 @@ main (int argc, char **argv)
   link_info.gc_sym_list = NULL;
   link_info.base_file = NULL;
   link_info.emit_hash = TRUE;
-  link_info.emit_gnu_hash = FALSE;
+#ifndef __mips__
+  link_info.emit_gnu_hash = TRUE;
+#endif
   /* SVR4 linkers seem to set DT_INIT and DT_FINI based on magic _init
      and _fini symbols.  We are compatible.  */
   link_info.init_function = "_init";
