--- ./binutils/configure.mps	2004-03-11 11:36:17.000000000 +0100
+++ ./binutils/configure	2004-03-11 11:35:26.000000000 +0100
@@ -1573,6 +1573,11 @@
   esac
   lt_cv_file_magic_test_file=`echo /lib/libc.so* /lib/libc-*.so`
   ;;
+  
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
 
 netbsd*)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
--- ./gprof/configure.mps	2004-03-11 11:20:54.000000000 +0100
+++ ./gprof/configure	2004-03-11 11:33:26.000000000 +0100
@@ -1569,6 +1569,11 @@
   esac
   lt_cv_file_magic_test_file=`echo /lib/libc.so* /lib/libc-*.so`
   ;;
+  
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
 
 netbsd*)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
--- ./opcodes/configure.mps	2004-03-11 11:32:15.000000000 +0100
+++ ./opcodes/configure	2004-03-11 11:31:43.000000000 +0100
@@ -1688,6 +1688,11 @@
   esac
   lt_cv_file_magic_test_file=`echo /lib/libc.so* /lib/libc-*.so`
   ;;
+  
+linux-uclibc*)
+  lt_cv_deplibs_check_method=pass_all
+  lt_cv_file_magic_test_file=`echo /lib/libuClibc-*.so`
+  ;;
 
 netbsd*)
   if echo __ELF__ | $CC -E - | grep __ELF__ > /dev/null; then
