--- binutils/ld/ldmain.c
+++ binutils/ld/ldmain.c
@@ -306,6 +306,7 @@ main (int argc, char **argv)
   link_info.strip_discarded = TRUE;
   link_info.callbacks = &link_callbacks;
   link_info.emit_hash = TRUE;
+  link_info.emit_gnu_hash = TRUE;
   /* SVR4 linkers seem to set DT_INIT and DT_FINI based on magic _init
      and _fini symbols.  We are compatible.  */
   link_info.init_function = "_init";
