--- ./ld/ldlang.c.tbss~	Wed May 26 21:01:36 2004
+++ ./ld/ldlang.c	Wed May 26 21:11:05 2004
@@ -3424,9 +3424,11 @@
 		dot = os->bfd_section->vma;
 		(void) lang_do_assignments_1 (os->children.head, os,
 					      os->fill, dot);
-		dot = (os->bfd_section->vma
-		       + TO_ADDR (os->bfd_section->_raw_size));
-
+		/* .tbss sections have zero size.  */
+		if ((os->bfd_section->flags & SEC_HAS_CONTENTS) != 0
+		    || (os->bfd_section->flags & SEC_THREAD_LOCAL) == 0
+		    || link_info.relocatable)
+		  dot += TO_ADDR (os->bfd_section->_raw_size);
 	      }
 	    if (os->load_base)
 	      {
