Index: gas/testsuite/gas/i386/i386.exp
===================================================================
RCS file: /cvs/src/src/gas/testsuite/gas/i386/i386.exp,v
retrieving revision 1.20
retrieving revision 1.20.6.1
diff -u -p -r1.20 -r1.20.6.1
--- gas/testsuite/gas/i386/i386.exp	22 Nov 2003 04:47:17 -0000	1.20
+++ gas/testsuite/gas/i386/i386.exp	12 Mar 2004 13:44:59 -0000	1.20.6.1
@@ -58,6 +58,7 @@ if [expr ([istarget "i*86-*-*"] ||  [ist
     run_dump_test "sub"
     run_dump_test "prescott"
     run_dump_test "divide"
+    run_dump_test "padlock"
 
     # PIC is only supported on ELF targets.
     if { ([istarget "*-*-elf*"] || [istarget "*-*-linux*"] )
Index: gas/testsuite/gas/i386/padlock.d
===================================================================
RCS file: gas/testsuite/gas/i386/padlock.d
diff -N gas/testsuite/gas/i386/padlock.d
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ gas/testsuite/gas/i386/padlock.d	15 Mar 2004 01:08:24 -0000	1.1.2.2
@@ -0,0 +1,21 @@
+#objdump: -dw
+#name: i386 padlock
+
+.*: +file format .*
+
+Disassembly of section .text:
+
+00000000 <foo>:
+   0:[	 ]*0f a7 c0 [	 ]*xstorerng 
+   3:[	 ]*f3 0f a7 c0 [	 ]*repz xstorerng 
+   7:[	 ]*f3 0f a7 c8 [	 ]*repz xcryptecb 
+   b:[	 ]*f3 0f a7 c8 [	 ]*repz xcryptecb 
+   f:[	 ]*f3 0f a7 d0 [	 ]*repz xcryptcbc 
+  13:[	 ]*f3 0f a7 d0 [	 ]*repz xcryptcbc 
+  17:[	 ]*f3 0f a7 e0 [	 ]*repz xcryptcfb 
+  1b:[	 ]*f3 0f a7 e0 [	 ]*repz xcryptcfb 
+  1f:[	 ]*f3 0f a7 e8 [	 ]*repz xcryptofb 
+  23:[	 ]*f3 0f a7 e8 [	 ]*repz xcryptofb 
+  27:[	 ]*0f a7 c0 [	 ]*xstorerng 
+  2a:[	 ]*f3 0f a7 c0 [	 ]*repz xstorerng 
+[ 	]*\.\.\.
Index: gas/testsuite/gas/i386/padlock.s
===================================================================
RCS file: gas/testsuite/gas/i386/padlock.s
diff -N gas/testsuite/gas/i386/padlock.s
--- /dev/null	1 Jan 1970 00:00:00 -0000
+++ gas/testsuite/gas/i386/padlock.s	15 Mar 2004 01:08:24 -0000	1.1.2.2
@@ -0,0 +1,18 @@
+# VIA Nehemiah PadLock instructions
+
+	.text
+foo:
+	xstorerng
+	rep xstorerng
+	xcryptecb
+	rep xcryptecb
+	xcryptcbc
+	rep xcryptcbc
+	xcryptcfb
+	rep xcryptcfb
+	xcryptofb
+	rep xcryptofb
+	xstore
+	rep xstore
+
+	.p2align 4,0
