--- binutils-2.15.90.0.1.1.old/ld/ldlang.c	2004-05-10 03:24:18.200798584 -0400
+++ binutils-2.15.90.0.1.1/ld/ldlang.c	2004-05-10 03:26:59.645255296 -0400
@@ -3453,9 +3453,11 @@
 		dot = os->bfd_section->vma;
 		(void) lang_do_assignments_1 (os->children.head, os,
 					      os->fill, dot);
-		dot = (os->bfd_section->vma
-		       + TO_ADDR (os->bfd_section->_raw_size));
-
+ 		/* .tbss sections effectively have zero size.  */
+ 		if ((os->bfd_section->flags & SEC_HAS_CONTENTS) != 0
+ 		    || (os->bfd_section->flags & SEC_THREAD_LOCAL) == 0
+ 		    || link_info.relocatable)
+ 		  dot += TO_ADDR (os->bfd_section->_raw_size);
 	      }
 	    if (os->load_base)
 	      {
