--- bfd/elf32-ppc.c
+++ bfd/elf32-ppc.c	2003/04/24 17:09:57
@@ -4944,6 +4944,15 @@ ppc_elf_relocate_section (output_bfd, in
 	  goto dodyn;
 
 	case R_PPC_REL24:
+	  if (info->shared) {
+		  (*_bfd_error_handler)(_("Error: The symbol `%s' has a %s relocation, that means '%s' was compiled without -fPIC."),
+		  	sym_name,
+			howto->name,
+			bfd_archive_filename (input_bfd));
+#if 1
+		  ret = FALSE;
+#endif
+	  }
 	case R_PPC_REL32:
 	case R_PPC_REL14:
 	case R_PPC_REL14_BRTAKEN:
