--- binutils/ld/ldmain.c
+++ binutils/ld/ldmain.c
@@ -273,2 +273,5 @@ main (int argc, char **argv)
   link_info.emit_hash = TRUE;
+#ifndef __mips__
+  link_info.emit_gnu_hash = TRUE;
+#endif
   link_info.callbacks = &link_callbacks;

work around minor linker bug:
http://sourceware.org/bugzilla/show_bug.cgi?id=11355

--- binutils/ld/testsuite//ld-elfvsb/hidden2.d
+++ binutils/ld/testsuite//ld-elfvsb/hidden2.d
@@ -1,5 +1,5 @@
 #source: hidden2.s
-#ld: -shared -T hidden2.ld
+#ld: -shared -T hidden2.ld --hash-style=sysv
 #readelf: -Ds
 # It is also ok to remove this symbol, but we currently make it local.
 
